### DISCUSSION TITLE
Ingame scripting and microcontrollers

### DISCUSSION THREAD

**Discussion by CapnMachaddish** [FakeFishGames/Barotrauma#11488] [Ideas and suggestions]
Players have been making hugely complicated systems of automation with wiring components for ages, but it's gotten to where it's commonplace, to the point of being almost _standard practice_, to have massive "walls" of wiring outside the sub that are all hidden ingame. As much as I like all of the fantastic automation and gadgets these enable, I feel like it's a bit ridiculous to have so much stuff that's entirely _out of the hands of the crew while in-round_. A lot of sandbox games with wiring-type systems include a form of very basic ingame scripting in the form of some kind of "microcontroller component" or such (for example, Space Engineers' Programmable Block, or Minecraft's Computercraft mod).

I think Barotrauma would greatly benefit from such a feature as well - partially mitigating wiring clutter and opening up more avenues for dynamic shenanigans in-round.

I'd love to see Terminals gain functionality as such microcontrollers, where players and sub creators can write simple scripts that can read input and write output to the Terminal's wire connections and text interface. 
A related possibility could also be an actual microcontroller _component_, which has no Terminal text I/O but can still run simple scripts operating on its wire inputs and outputs.

**Comment by zhu-rengong** [comment]
I totally agree with your idea, In multiplayer mode, many circuits will occupy a lot of the game's computing resources, and it is almost impossible to have more than a dozen submarines existing and running smoothly on one map. This is also the main reason why I wrote programmable components [(LuaNet Component)](https://steamcommunity.com/sharedfiles/filedetails/?id=2937571024)
